oktapam/resource_ad_user_sync_task_settings.go

```diff
@@ -0,0 +1,231 @@
+package oktapam
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/okta/terraform-provider-oktapam/oktapam/client"
+	"github.com/okta/terraform-provider-oktapam/oktapam/constants/attributes"
+	"github.com/okta/terraform-provider-oktapam/oktapam/constants/descriptions"
+	"github.com/okta/terraform-provider-oktapam/oktapam/logging"
+	"github.com/okta/terraform-provider-oktapam/oktapam/utils"
+)
+
+func resourceADUserSyncTaskSettings() *schema.Resource {
+	/** ADUserSyncTaskSettings is immutable. Only updatable attributes are active status and task frequency/start time.
+		ADUserSyncTaskSettings resource has update endpoint, but internally it's a transactional delete & then create
+	    operation. Terraform resource id change and deletion of existing resource confuse terraform. Not sure if there is
+	    any way to handle it properly. For now, I have marked all the immutable attributes as "ForceNew". If update
+	    ADUserSyncTaskSettings endpoint has additional logic then we need to revisit this.
+	    TF Reference - https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors#forcenew
+	*/
+	return &schema.Resource{
+		Description:   descriptions.ResourceADUserSyncTaskSettings,
+		CreateContext: resourceADUserSyncTaskSettingsCreate,
+		ReadContext:   resourceADUserSyncTaskSettingsRead,
+		UpdateContext: resourceADUserSyncTaskSettingsUpdate,
+		DeleteContext: resourceADUserSyncTaskSettingsDelete,
+		Schema: map[string]*schema.Schema{
+			attributes.ADConnectionID: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: descriptions.ADConnectionID,
+			},
+			attributes.Name: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.Name,
+			},
+			attributes.ID: {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			attributes.Frequency: {
+				Type:         schema.TypeInt,
+				Required:     true,
+				ValidateFunc: validation.IntBetween(1, 24),
+				Description:  descriptions.ADUserSyncTaskSettingsFrequency,
+			},
+			attributes.StartHourUTC: {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(0, 23),
+				Description:  descriptions.ADUserSyncTaskSettingsStartHourUTC,
+			},
+			attributes.BaseDN: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.ADUserSyncTaskSettingsBaseDN,
+			},
+			attributes.LDAPQueryFilter: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.ADUserSyncTaskSettingsLDAPQueryFilter,
+			},
+			attributes.UPNField: {
```

Comment:
This field should be marked as optional. I think at database end we can set default. Check how it's handled for AdTaskSettings -> HostNameAttribute

Reply:
Task settings use those as Required/Forcenew, but may be because there the API does not have default. I can try optional here, we have to deal with the update semantics, I guess we can say missing attribute there means use default (and not previous values).

Reply:
ok